SM panics after restart after cluster deletion

[PR](https://github.com/scylladb/scylla-manager/pull/4466) adding control over healtcheck intervals introduced regression detected in [SCT](https://argus.scylladb.com/tests/scylla-cluster-tests/b1fefe20-98a2-4a92-b275-fb646477c471/events):
```
panic: unknown healthcheck mode:
goroutine 1 [running]:
github.com/scylladb/scylla-manager/v3/pkg/service/scheduler.(*Service).UpdateHealthcheckTasks.func1(0xc00058d620?)
```
It is caused by the following part of the code:
```go
// UpdateHealthcheckTasks should be called on start before LoadTasks.
// It updates healthcheck tasks cron schedule to match the one specified in the config.
func (s *Service) UpdateHealthcheckTasks(ctx context.Context, cfg healthcheck.Config) error {
	err := s.forEachTask(func(t *Task) error {
		if t.Type != HealthCheckTask {
			return nil
		}
		var expected schedules.Cron
		switch healthcheck.Mode(t.Name) {
		case healthcheck.CQLMode:
			expected = cfg.CQLPingCron
		case healthcheck.RESTMode:
			expected = cfg.RESTPingCron
		case healthcheck.AlternatorMode:
			expected = cfg.AlternatorPingCron
		default:
			panic("unknown healthcheck mode: " + t.Name)
		}
		if expected.CronSpecification != t.Sched.Cron.CronSpecification {
			s.logger.Info(ctx, "Updating healthcheck task cron schedule", "mode", t.Name, "task ID", t.ID)
			t.Sched.Cron = expected
			return s.PutTask(ctx, t)
		}
		return nil
	})
	return errors.Wrap(err, "iterate over tasks")
}
```

In this code, healtcheck tasks are differentiated by their names.
The problem is that when we delete task or an entire cluster (which causes task deletions), we are leaving task definition in SM DB, but we clean its name:
```go
func (s *Service) DeleteTask(ctx context.Context, t *Task) error {
	s.logger.Debug(ctx, "DeleteTask", "task", t)

	t.Deleted = true
	t.Enabled = false

	// Remove the deleted task's name so that new tasks can use it
	t.Name = ""
```
This shows that relying on healthcheck task names for differentiating healthcheck tasks is fragile and should be fixed.
Luckily for us, this regression hasn't made it to any release, so users were affected.

cc: @mikliapko

--- a/pkg/cmd/scylla-manager/server.go
+++ b/pkg/cmd/scylla-manager/server.go
@@ -352,8 +352,10 @@ func (s *server) startServices(ctx context.Context) error {
 	if err := s.clusterSvc.Init(ctx); err != nil {
 		return errors.Wrap(err, "cluster service")
 	}
+	// Not being able to update healthcheck tasks is not critical,
+	// so we shouldn't fail service start because of that.
 	if err := s.schedSvc.UpdateHealthcheckTasks(ctx, s.config.Healthcheck); err != nil {
-		return errors.Wrap(err, "update healthcheck tasks")
+		s.logger.Error(ctx, "Failed to update healthcheck tasks", "error", err)
 	}
 	if err := s.schedSvc.LoadTasks(ctx); err != nil {
 		return errors.Wrap(err, "schedule service")

--- a/pkg/service/healthcheck/runner.go
+++ b/pkg/service/healthcheck/runner.go
@@ -27,12 +27,12 @@ type Runner struct {
 }
 
 func (r Runner) Run(ctx context.Context, clusterID, taskID, runID uuid.UUID, properties json.RawMessage) error {
-	p := taskProperties{}
-	if err := json.Unmarshal(properties, &p); err != nil {
-		return util.ErrValidate(err)
+	m, err := ModeFromProperties(properties)
+	if err != nil {
+		return err
 	}
 
-	switch p.Mode {
+	switch m {
 	case CQLMode:
 		return r.cql.Run(ctx, clusterID, taskID, runID, properties)
 	case RESTMode:
@@ -42,6 +42,15 @@ func (r Runner) Run(ctx context.Context, clusterID, taskID, runID uuid.UUID, pro
 	default:
 		return errors.Errorf("unspecified mode")
 	}
+}
+
+// ModeFromProperties return Mode of healthcheck task based on its properties.
+func ModeFromProperties(properties json.RawMessage) (Mode, error) {
+	p := taskProperties{}
+	if err := json.Unmarshal(properties, &p); err != nil {
+		return "", util.ErrValidate(err)
+	}
+	return p.Mode, nil
 }
 
 type runner struct {

--- a/pkg/service/scheduler/service.go
+++ b/pkg/service/scheduler/service.go
@@ -5,6 +5,7 @@ package scheduler
 import (
 	"context"
 	"encoding/json"
+	stdErr "errors"
 	"sync"
 	"time"
 	"unsafe"
@@ -125,13 +126,20 @@ func (s *Service) mustRunner(tp TaskType) Runner {
 
 // UpdateHealthcheckTasks should be called on start before LoadTasks.
 // It updates healthcheck tasks cron schedule to match the one specified in the config.
+// It tries to process all tasks before returning any error.
 func (s *Service) UpdateHealthcheckTasks(ctx context.Context, cfg healthcheck.Config) error {
+	var iterErr error
 	err := s.forEachTask(func(t *Task) error {
-		if t.Type != HealthCheckTask {
+		if t.Type != HealthCheckTask || t.Deleted {
+			return nil
+		}
+		m, err := healthcheck.ModeFromProperties(t.Properties)
+		if err != nil {
+			iterErr = stdErr.Join(iterErr, errors.Wrapf(err, "task %s: get healthcheck task mode", t.ID))
 			return nil
 		}
 		var expected schedules.Cron
-		switch healthcheck.Mode(t.Name) {
+		switch m {
 		case healthcheck.CQLMode:
 			expected = cfg.CQLPingCron
 		case healthcheck.RESTMode:
@@ -139,16 +147,20 @@ func (s *Service) UpdateHealthcheckTasks(ctx context.Context, cfg healthcheck.Co
 		case healthcheck.AlternatorMode:
 			expected = cfg.AlternatorPingCron
 		default:
-			panic("unknown healthcheck mode: " + t.Name)
+			iterErr = stdErr.Join(iterErr, errors.Errorf("task %s: unknown healthcheck mode: %s", t.ID, m))
+			return nil
 		}
 		if expected.CronSpecification != t.Sched.Cron.CronSpecification {
 			s.logger.Info(ctx, "Updating healthcheck task cron schedule", "mode", t.Name, "task ID", t.ID)
 			t.Sched.Cron = expected
-			return s.PutTask(ctx, t)
+			if err := s.PutTask(ctx, t); err != nil {
+				iterErr = stdErr.Join(iterErr, errors.Wrapf(err, "task %s: put task", t.ID))
+				return nil
+			}
 		}
 		return nil
 	})
-	return errors.Wrap(err, "iterate over tasks")
+	return stdErr.Join(iterErr, errors.Wrap(err, "iterate over tasks"))
 }
 
 // LoadTasks should be called on start it loads and schedules task from database.

--- a/pkg/service/scheduler/service_integration_test.go
+++ b/pkg/service/scheduler/service_integration_test.go
@@ -341,6 +341,7 @@ func TestServiceScheduleIntegration(t *testing.T) {
 				Sched: scheduler.Schedule{
 					Cron: healthcheck.DefaultConfig().CQLPingCron,
 				},
+				Properties: json.RawMessage(`{"mode": "cql"}`),
 			},
 			{
 				ClusterID: h.clusterID,
@@ -350,6 +351,7 @@ func TestServiceScheduleIntegration(t *testing.T) {
 				Sched: scheduler.Schedule{
 					Cron: healthcheck.DefaultConfig().RESTPingCron,
 				},
+				Properties: json.RawMessage(`{"mode": "rest"}`),
 			},
 			{
 				ClusterID: h.clusterID,
@@ -359,6 +361,7 @@ func TestServiceScheduleIntegration(t *testing.T) {
 				Sched: scheduler.Schedule{
 					Cron: healthcheck.DefaultConfig().AlternatorPingCron,
 				},
+				Properties: json.RawMessage(`{"mode": "alternator"}`),
 			},
 		}
 		for _, task := range tasks {
@@ -407,6 +410,17 @@ func TestServiceScheduleIntegration(t *testing.T) {
 			if nameToSpec[task.Name] != task.Sched.Cron.Spec {
 				t.Fatalf("Healthcheck task %s cron spec %q: expected %q", task.Name, task.Sched.Cron.Spec, nameToSpec[task.Name])
 			}
+		}
+
+		// Checks for #4599
+		Print("When: delete one healthcheck task")
+		if err := h.service.DeleteTask(ctx, tasks[0]); err != nil {
+			t.Fatal(err)
+		}
+
+		Print("Then: update healthcheck tasks succeeds")
+		if err := h.service.UpdateHealthcheckTasks(ctx, cfg); err != nil {
+			t.Fatal(err)
 		}
 	})
 

